src/os/bluestore/BlueFS.cc

```diff
@@ -1448,9 +1451,12 @@ int BlueFS::_read(
     outbl->clear();
 
   int ret = 0;
+  std::shared_lock s_lock(h->lock);
   while (len > 0) {
     size_t left;
     if (off < buf->bl_off || off >= buf->get_buf_end()) {
+      s_lock.unlock();
+      std::unique_lock u_lock(h->lock);
```

Comment:
this looks wrong.  if we drop and retake the lock, don't we need to revalidate the conditions checked above?

Reply:
You're right. See https://github.com/ceph/ceph/pull/29012 for the fix.

Reply:
@liewegas - fix has been cherry-picked from #29012